2.0/m_extbanbanlist.cpp

```diff
@@ -0,0 +1,129 @@
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* $ModAuthor: genius3000 */
+/* $ModAuthorMail: genius3000@g3k.solutions */
+/* $ModDesc: Provides extban 'b' - Ban list from another channel */
+/* $ModDepends: core 2.0 */
+
+/* Helpop Lines for the EXTBANS section
+ * Find: '<helpop key="extbans" value="Extended Bans'
+ * Place just before the 'j:<channel>' line:
+ b:<channel>   Matches users banned in the given channel (requires
+               extbanbanlist extras-module).
+ */
+
+#include "inspircd.h"
+
+
+class ExtbanBanlist : public ModeWatcher
+{
+ public:
+	ExtbanBanlist(Module* parent) : ModeWatcher(parent, 'b', MODETYPE_CHANNEL)
+	{
+	}
+
+	bool BeforeMode(User* source, User* dest, Channel* channel, std::string& param, bool adding, ModeType type)
+	{
+		if (!IS_LOCAL(source) || type != MODETYPE_CHANNEL || !channel || !adding || param.length() < 3)
+			return true;
+
+		if (param[0] != 'b' || param[1] != ':')
+			return true;
+
+		Channel* chan = ServerInstance->FindChan(param.substr(2));
+		if (!chan)
+		{
+			source->WriteNumeric(ERR_NOSUCHCHANNEL, "%s %s :No such channel", source->nick.c_str(), param.substr(2).c_str());
+			return false;
+		}
+
+		if (chan == channel)
+		{
+			source->WriteNumeric(ERR_NOSUCHCHANNEL, "%s %s :Target channel must be a different channel", source->nick.c_str(), param.substr(2).c_str());
+			return false;
+		}
+
+		ModeHandler* mh = ServerInstance->Modes->FindMode('b', MODETYPE_CHANNEL);
+		if (mh->GetLevelRequired() > chan->GetPrefixValue(source))
+		{
+			source->WriteNumeric(ERR_CHANOPRIVSNEEDED, "%s %s :You must have access to modify the banlist to use it", source->nick.c_str(), param.substr(2).c_str());
+			return false;
+		}
+		return true;
+	}
+};
+
+class ModuleExtbanBanlist : public Module
+{
+	ExtbanBanlist eb;
+
+ public:
+	ModuleExtbanBanlist() : eb(this)
+	{
+	}
+
+	void init()
+	{
+		if (!ServerInstance->Modes->AddModeWatcher(&eb))
+			throw ModuleException("Could not add mode watcher");
+
+		Implementation eventlist[] = { I_OnCheckBan, I_On005Numeric };
+		ServerInstance->Modules->Attach(eventlist, this, sizeof(eventlist)/sizeof(Implementation));
+	}
+
+	~ModuleExtbanBanlist()
+	{
+		ServerInstance->Modes->DelModeWatcher(&eb);
+	}
+
+	ModResult OnCheckBan(User* user, Channel* c, const std::string& mask)
+	{
+		if ((mask.length() > 2) && (mask[0] == 'b') && (mask[1] == ':'))
+		{
+			Channel* chan = ServerInstance->FindChan(mask.substr(2));
+			if (!chan)
+				return MOD_RES_PASSTHRU;
+
+			for (BanList::iterator ban = chan->bans.begin(); ban != chan->bans.end(); ++ban)
+			{
+				std::string& banmask = ban->data;
+
+				// Prevent multiple (and circular) banlist loops
+				if (banmask.length() > 2 && banmask[0] == 'b' && banmask[1] == ':')
```

Comment:
How does this handle `U:b:#channel`? From my understanding, that would cause it to loop. Possibly setting a module-wide boolean while iterating the banlist and doing something like 
```c++
if (this->checking)
    return MOD_RES_PASSTHRU;
```
at the beginning of `OnCheckBan()` would help prevent odd cases like this.

If the above change is made, the for-loop can be replaced with a call to `chan->IsBanned(user)` which would also take `+e` in to consideration.

What I'm thinking is something like
```c++
ModResult OnCheckBan(User* user, Channel* c, const std::string& mask)
{
    if ((mask.length() > 2) && (mask[0] == 'b') && (mask[1] == ':') && !checking)
    {
        Channel* chan = ServerInstance->FindChan(mask.substr(2));
        if (!chan)
            return MOD_RES_PASSTHRU;

        this->checking = true;
        bool matched = chan->IsBanned(user);
        this->checking = false;
        if (matched)
            return MOD_RES_DENY;
    }
    return MOD_RES_PASSTHRU;
}
```

Reply:
I don't recall if I tested such a situation but just going from memory it wouldn't get in an endless loop. All ExtBans run a check on mask.substr(2) (or similar) which will still end up with this module receiving "b:#channel" and it will stop there.  
I'm not sure I want to check exceptions. I'm open to community feedback on this, but personally I see including another channel's banlist as just that, nothing else.

Reply:
`MODE #channel +b U:b:#channel1`
`MODE #channel1 +b U:b:#channel`
After this, it will reach an infinite loop.

Reply:
Hmmm yeah I see your point here. Although, upon testing I'm not seeing any issues, which is strange. Either way, I will rework it for better safety.